README.md
fork-dbfpy
==========

I adapted dbfpy tool for my project fixing some things maybe are not correct in a regular dbf file, or maybe yes, i don't know. if you feel confortable with this used that


See an example for read tables with this fork ``https://github.com/paridin/Migrating-Tool-from-dBase``


CHANGES
Version 2.3.0 (29-jun-2014)
---------------------------

Support Memo files.

Support getting a slice of a dbf object.

Version 2.2.5 (16-sep-2010)
---------------------------

Fix Y2K issue with Last Update field in the header (sf bug 3065838).

Version 2.2.4 (26-may-2009)
---------------------------

Ignore leading and trailing zero bytes in numeric fields.

Version 2.2.3 (17-feb-2009)
---------------------------

Support for writing empty date values.

Version 2.2.2 (16-sep-2008)
---------------------------

Numeric decoder returns float when the value contains decimal point
without regard to the number of decimal places declared in the header.

Version 2.2.1 (16-mar-2008)
---------------------------

Fix: raise ValueError if a name passed to field constructor
is longer than 10 characters.

Version 2.2.0 (11-feb-2007)
---------------------------

Features:
  - date fields are allowed to contain values stored with leading spaces
    instead of leading zeroes.
  - dbf.header returns field definition objects if accessed as a list
    or as a dictionary.
  - added raw data access methods: DbfRecord.rawFromStream(),
    DbfFieldDef.rawFromRecord().
  - added conversion error handling: if ignoreErrors=True is passed
    with Dbf constructor arguments, then failing field value conversions
    will not raise errors but instead will return special object
    INVALID_VALUE which is equal to None, empty string and zero.
    .ignoreErrors property of the Dbf instances may be toggled also
    after instance initialization.

Version 2.1.0 (01-dec-2006)
---------------------------

Features:
  - support field types 'F' (float), 'I' (integer) and 'Y' (currency)

Fix processing of empty Timestamp values

Version 2.0.3 (30-oct-2006)
---------------------------

Fixes:
  - compatibility fix for Python versions prior to 2.4 (sf bug 1574526)
  - wrong record length when reading from file object (sf bug 1586619)

Version 2.0.2 (08-jul-2006)
---------------------------

Fix dbfnew (legacy API for DBF creation)

Version 2.0.1 (10-mar-2006)
---------------------------

Fixes in Numeric field processing:
  - decoding inserted decimal point where it shouldn't;
  - encoded value never exceeds field length.
    If encoded value is too large for the field, decimal digits get
    removed (rounding the value down).  When integer value is larger
    than the field length, ValueError is raised.

Fix: Date and Logical fields were not able to decode empty values.

Errors raised from field processing include field name.

This file is included in the source distribution.

Version 2.0.0 (20-dec-2005)
---------------------------

Changed style of doc-strings; added more docs.

``mx.DateTime`` isn't required anymore - internal representation uses
``datetime.date`` and ``datetime.datetime`` classes from the batteries.

'D' and 'T' became "smarter":
  - date fields could accept:
    - date/time tuples [or tuples with at least
      first three components (year, day, month)];
    - strings in format 'yyyymmdd' or 'yymmdd';
    - ``datetime.date`` and ``datetime.datetime`` instances;
    - ``None`` instance - use current date value;
    - numbers (``int``, ``long``, ``float``)
      which are treated as seconds from epoch;
    - objects having 'ticks' method (``mx.DateTime`` for example);
  - datetime fields can accept everything date fields do
    except string objects (which is a TODO stuff).

Changed identifiers and args of some methods. For example old code

         dbf = Dbf()
         dbf.openFile("county.dbf", readOnly=1)

now should be written like

         dbf = Dbf("county.dbf", True)

All classes now are new-style classes (inherited from the ``object``).

Got rid of ``binnum`` and ``strutil`` modules,
using ``struct`` and string methods instead.

Group classes in different modules instead of having them all in
one, as it was earlier (see dbf.py from the previous release(s)).

Use fields' registry to simplify registering of the new data types.


Previous changes:
-----------------

took over development 2000-10-06 Hans Fiby

changes: 2001-02-07 Hans Fiby
 * read long integers as long
 * add this CHANGES File

changed: 2000-10-06 Hans Fiby
 * added dbfnew.py
 * fixed decimal points


dbfpy/fields.py
"""DBF fields definitions.

TODO:
  - make memos work
"""
"""History (most recent first):
14-dec-2010 [als]   support reading and writing Memo fields;
                    .toString: write field offset
26-may-2009 [als]   DbfNumericFieldDef.decodeValue: strip zero bytes
05-feb-2009 [als]   DbfDateFieldDef.encodeValue: empty arg produces empty date
16-sep-2008 [als]   DbfNumericFieldDef decoding looks for decimal point
                    in the value to select float or integer return type
13-mar-2008 [als]   check field name length in constructor
11-feb-2007 [als]   handle value conversion errors
10-feb-2007 [als]   DbfFieldDef: added .rawFromRecord()
01-dec-2006 [als]   Timestamp columns use None for empty values
31-oct-2006 [als]   support field types 'F' (float), 'I' (integer)
                    and 'Y' (currency);
                    automate export and registration of field classes
04-jul-2006 [als]   added export declaration
10-mar-2006 [als]   decode empty values for Date and Logical fields;
                    show field name in errors
10-mar-2006 [als]   fix Numeric value decoding: according to spec,
                    value always is string representation of the number;
                    ensure that encoded Numeric value fits into the field
20-dec-2005 [yc]    use field names in upper case
15-dec-2005 [yc]    field definitions moved from `dbf`.
"""

__version__ = "$Revision: 1.15 $"[11:-2]
__date__ = "$Date: 2010/12/14 11:04:49 $"[7:-2]

__all__ = ["lookupFor",] # field classes added at the end of the module

import datetime
import struct
import sys

from memo import MemoData
import utils

## abstract definitions

class DbfFieldDef(object):
    """Abstract field definition.

    Child classes must override ``type`` class attribute to provide datatype
    infromation of the field definition. For more info about types visit
    `http://www.clicketyclick.dk/databases/xbase/format/data_types.html`

    Also child classes must override ``defaultValue`` field to provide
    default value for the field value.

    If child class has fixed length ``length`` class attribute must be
    overriden and set to the valid value. None value means, that field
    isn't of fixed length.

    Note: ``name`` field must not be changed after instantiation.

    """

    __slots__ = ("name", "length", "decimalCount",
        "start", "end", "ignoreErrors")

    # length of the field, None in case of variable-length field,
    # or a number if this field is a fixed-length field
    length = None

    # field type. for more information about fields types visit
    # `http://www.clicketyclick.dk/databases/xbase/format/data_types.html`
    # must be overriden in child classes
    typeCode = None

    # default value for the field. this field must be
    # overriden in child classes
    defaultValue = None

    # True if field data is kept in the Memo file
    isMemo = property(lambda self: self.typeCode in "GMP")

    def __init__(self, name, length=None, decimalCount=None,
        start=None, stop=None, ignoreErrors=False,
    ):
        """Initialize instance."""
        assert self.typeCode is not None, "Type code must be overriden"
        assert self.defaultValue is not None, "Default value must be overriden"
        ## fix arguments
        if len(name) >10:
            raise ValueError("Field name \"%s\" is too long" % name)
        name = str(name).upper()
        if self.__class__.length is None:
            if length is None:
                raise ValueError("[%s] Length isn't specified" % name)
            length = int(length)
            if length <= 0:
                raise ValueError("[%s] Length must be a positive integer"
                    % name)
        else:
            length = self.length
        if decimalCount is None:
            decimalCount = 0
        ## set fields
        self.name = name
        # FIXME: validate length according to the specification at
        # http://www.clicketyclick.dk/databases/xbase/format/data_types.html
        self.length = length
        self.decimalCount = decimalCount
        self.ignoreErrors = ignoreErrors
        self.start = start
        self.end = stop

    def __cmp__(self, other):
        return cmp(self.name, str(other).upper())

    def __hash__(self):
        return hash(self.name)

    def fromString(cls, string, start, ignoreErrors=False):
        """Decode dbf field definition from the string data.

        Arguments:
            string:
                a string, dbf definition is decoded from. length of
                the string must be 32 bytes.
            start:
                position in the database file.
            ignoreErrors:
                initial error processing mode for the new field (boolean)

        """
        assert len(string) == 32
        _length = ord(string[16])
        return cls(utils.unzfill(string)[:11], _length, ord(string[17]),
            start, start + _length, ignoreErrors=ignoreErrors)
    fromString = classmethod(fromString)

    def toString(self):
        """Return encoded field definition.

        Return:
            Return value is a string object containing encoded
            definition of this field.

        """
        if sys.version_info < (2, 4):
            # earlier versions did not support padding character
            _name = self.name[:11] + "\0" * (11 - len(self.name))
        else:
            _name = self.name.ljust(11, '\0')
        return (
            _name +
            self.typeCode +
            struct.pack("<L", self.start) +
            chr(self.length) +
            chr(self.decimalCount) +
            chr(0) * 14
        )

    def __repr__(self):
        return "%-10s %1s %3d %3d" % self.fieldInfo()

    def fieldInfo(self):
        """Return field information.

        Return:
            Return value is a (name, type, length, decimals) tuple.

        """
        return (self.name, self.typeCode, self.length, self.decimalCount)

    def rawFromRecord(self, record):
        """Return a "raw" field value from the record string."""
        return record[self.start:self.end]

    def decodeFromRecord(self, record):
        """Return decoded field value from the record string."""
        try:
            return self.decodeValue(self.rawFromRecord(record))
        except:
            if self.ignoreErrors:
                return utils.INVALID_VALUE
            else:
                raise

    def decodeValue(self, value):
        """Return decoded value from string value.

        This method shouldn't be used publicly. It's called from the
        `decodeFromRecord` method.

        This is an abstract method and it must be overridden in child classes.
        """
        raise NotImplementedError

    def encodeValue(self, value):
        """Return str object containing encoded field value.

        This is an abstract method and it must be overriden in child classes.
        """
        raise NotImplementedError

## real classes

class DbfCharacterFieldDef(DbfFieldDef):
    """Definition of the character field."""

    typeCode = "C"
    defaultValue = ""

    def decodeValue(self, value):
        """Return string object.

        Return value is a ``value`` argument with stripped right spaces.

        """
        return value.rstrip(" ")

    def encodeValue(self, value):
        """Return raw data string encoded from a ``value``."""
        return str(value)[:self.length].ljust(self.length)


class DbfNumericFieldDef(DbfFieldDef):
    """Definition of the numeric field."""

    typeCode = "N"
    # XXX: now I'm not sure it was a good idea to make a class field
    # `defaultValue` instead of a generic method as it was implemented
    # previously -- it's ok with all types except number, cuz
    # if self.decimalCount is 0, we should return 0 and 0.0 otherwise.
    defaultValue = 0

    def decodeValue(self, value):
        """Return a number decoded from ``value``.

        If decimals is zero, value will be decoded as an integer;
        or as a float otherwise.

        Return:
            Return value is a int (long) or float instance.

        """
        value = value.strip(" \0")
        if "." in value:
            # a float (has decimal separator)
            return float(value)
        elif value:
            # must be an integer
            import re # detect is not a number set 0 to fix problems
            if re.match(r'\D', value): #Added September 8 2014
                return int(value)
        else:
            return 0

    def encodeValue(self, value):
        """Return string containing encoded ``value``."""
        _rv = ("%*.*f" % (self.length, self.decimalCount, value))
        if len(_rv) > self.length:
            _ppos = _rv.find(".")
            if 0 <= _ppos <= self.length:
                _rv = _rv[:self.length]
            else:
                raise ValueError("[%s] Numeric overflow: %s (field width: %i)"
                    % (self.name, _rv, self.length))
        return _rv

class DbfFloatFieldDef(DbfNumericFieldDef):
    """Definition of the float field - same as numeric."""

    typeCode = "F"

class DbfIntegerFieldDef(DbfFieldDef):
    """Definition of the integer field."""

    typeCode = "I"
    length = 4
    defaultValue = 0

    def decodeValue(self, value):
        """Return an integer number decoded from ``value``."""
        return struct.unpack("<i", value)[0]

    def encodeValue(self, value):
        """Return string containing encoded ``value``."""
        return struct.pack("<i", int(value))

class DbfCurrencyFieldDef(DbfFieldDef):
    """Definition of the currency field."""

    typeCode = "Y"
    length = 8
    defaultValue = 0.0

    def decodeValue(self, value):
        """Return float number decoded from ``value``."""
        return struct.unpack("<q", value)[0] / 10000.

    def encodeValue(self, value):
        """Return string containing encoded ``value``."""
        return struct.pack("<q", round(value * 10000))

class DbfLogicalFieldDef(DbfFieldDef):
    """Definition of the logical field."""

    typeCode = "L"
    defaultValue = -1
    length = 1

    def decodeValue(self, value):
        """Return True, False or -1 decoded from ``value``."""
        # Note: value always is 1-char string
        if value == "?":
            return -1
        if value in "NnFf ":
            return False
        if value in "YyTt":
            return True
        raise ValueError("[%s] Invalid logical value %r" % (self.name, value))

    def encodeValue(self, value):
        """Return a character from the "TF?" set.

        Return:
            Return value is "T" if ``value`` is True
            "?" if value is -1 or False otherwise.

        """
        if value is True:
            return "T"
        if value == -1:
            return "?"
        return "F"


class DbfMemoFieldDef(DbfFieldDef):
    """Definition of the memo field."""

    typeCode = "M"
    defaultValue = "\0" * 4
    length = 4
    # MemoFile instance.  Must be set before reading or writing to the field.
    file = None
    # MemoData type for strings written to the memo file
    memoType = MemoData.TYPE_MEMO

    def decodeValue(self, value):
        """Return MemoData instance containing field data."""
        _block = struct.unpack("<L", value)[0]
        if _block:
            return self.file.read(_block)
        else:
            return MemoData("", self.memoType)

    def encodeValue(self, value):
        """Return raw data string encoded from a ``value``.

        Note: this is an internal method.

        """
        if value:
            return struct.pack("<L",
                self.file.write(MemoData(value, self.memoType)))
        else:
            return self.defaultValue


class DbfGeneralFieldDef(DbfFieldDef):
    """Definition of the general (OLE object) field."""

    typeCode = "G"
    memoType = MemoData.TYPE_OBJECT


class DbfDateFieldDef(DbfFieldDef):
    """Definition of the date field."""

    typeCode = "D"
    defaultValue = utils.classproperty(lambda cls: datetime.date.today())
    # "yyyymmdd" gives us 8 characters
    length = 8

    def decodeValue(self, value):
        """Return a ``datetime.date`` instance decoded from ``value``."""
        if value.strip():
            return utils.getDate(value)
        else:
            return None

    def encodeValue(self, value):
        """Return a string-encoded value.

        ``value`` argument should be a value suitable for the
        `utils.getDate` call.

        Return:
            Return value is a string in format "yyyymmdd".

        """
        if value:
            return utils.getDate(value).strftime("%Y%m%d")
        else:
            return " " * self.length


class DbfDateTimeFieldDef(DbfFieldDef):
    """Definition of the timestamp field."""

    # a difference between JDN (Julian Day Number)
    # and GDN (Gregorian Day Number). note, that GDN < JDN
    JDN_GDN_DIFF = 1721425
    typeCode = "T"
    defaultValue = utils.classproperty(lambda cls: datetime.datetime.now())
    # two 32-bits integers representing JDN and amount of
    # milliseconds respectively gives us 8 bytes.
    # note, that values must be encoded in LE byteorder.
    length = 8

    def decodeValue(self, value):
        """Return a `datetime.datetime` instance."""
        assert len(value) == self.length
        # LE byteorder
        _jdn, _msecs = struct.unpack("<2I", value)
        if _jdn >= 1:
            _rv = datetime.datetime.fromordinal(_jdn - self.JDN_GDN_DIFF)
            _rv += datetime.timedelta(0, _msecs / 1000.0)
        else:
            # empty date
            _rv = None
        return _rv

    def encodeValue(self, value):
        """Return a string-encoded ``value``."""
        if value:
            value = utils.getDateTime(value)
            # LE byteorder
            _rv = struct.pack("<2I", value.toordinal() + self.JDN_GDN_DIFF,
                (value.hour * 3600 + value.minute * 60 + value.second) * 1000)
        else:
            _rv = "\0" * self.length
        assert len(_rv) == self.length
        return _rv


_fieldsRegistry = {}

def registerField(fieldCls):
    """Register field definition class.

    ``fieldCls`` should be subclass of the `DbfFieldDef`.

    Use `lookupFor` to retrieve field definition class
    by the type code.

    """
    assert fieldCls.typeCode is not None, "Type code isn't defined"
    # XXX: use fieldCls.typeCode.upper()? in case of any decign
    # don't forget to look to the same comment in ``lookupFor`` method
    _fieldsRegistry[fieldCls.typeCode] = fieldCls


def lookupFor(typeCode):
    """Return field definition class for the given type code.

    ``typeCode`` must be a single character. That type should be
    previously registered.

    Use `registerField` to register new field class.

    Return:
        Return value is a subclass of the `DbfFieldDef`.

    """
    # XXX: use typeCode.upper()? in case of any decign don't
    # forget to look to the same comment in ``registerField``
    return _fieldsRegistry[typeCode]

## register generic types

for (_name, _val) in globals().items():
    if isinstance(_val, type) and issubclass(_val, DbfFieldDef) \
    and (_name != "DbfFieldDef"):
        __all__.append(_name)
        registerField(_val)
del _name, _val

# vim: et sts=4 sw=4 :
